### DISCUSSION TITLE
AWS Lambda / How to set Cookies inside Procedures

### DISCUSSION THREAD

**Discussion by aheissenberger** [trpc/trpc#4420] [Q&A]
How can I set and remove cookies similar to ctx.res.cookie(..) with Express adapter when using `@trpc/server/adapters/aws-lambda`?

My configuration is:
* node 18
* tRPC 10
* aws lambda adapter / APIGatewayProxyEventV2

I do not need CORS - I only need to read/write/delete Cookies

Thank you for the support and this great tool!
Andreas

**Comment by tgdn** [comment]
@aheissenberger did you figure it out?

**Reply by Nick-Lucas** [reply]
Only way really would be to include the response type in the Context, and then use that directly. But we wouldn't really recommend it as it couples your procedures to the host adapter, and tRPC isn't a HTTP/REST framework so setting cookies natively isn't really in scope.

Context is how you'd do it if you really want though

**Comment by anatolzak** [comment]
This is what I have been able to come up with:

It works ok for setting 1 cookie, but I encountered a weird typescript error.

When I include the `Set-Cookie` header as an array of strings in the `responseMeta`, TypeScript flags an error. This occurs because the type definitions expect headers to be an object with string values, not arrays. However, to set multiple cookies, browsers require that each cookie is sent in its own Set-Cookie header, which necessitates using an array of strings in this context.

So with the lambda adapter you can only set 1 cookie. However, you can use the same code and it will successfully set multiple cookies when using a standalone adapter.

The standalone adapter actually adds multiple `Set-Cookie` headers to the response. With the lambda adapter, it just joins the array of cookies into a string (i.e. `session_id=1,user_id=1`)

```typescript
// routers/auth.ts
export const auth = router({
  login: publicProcedure.mutation(({ ctx }) => {
    ctx.resp.cookie('session_id', crypto.randomBytes(128).toString('base64url'));

    return {
      message: 'Logged in successfully',
    };
  }),
  logout: publicProcedure.mutation(({ ctx }) => {
    ctx.resp.clearCookie('session_id');

    return {
      message: 'Logged out successfully',
    };
  }),
});
```

```typescript
// lambda.ts
export const handler = awsLambdaRequestHandler({
  router: appRouter,
  createContext,
  responseMeta: ({ ctx }) => {
    const headers: Record<string, string | string[]> = { ...ctx?.resp.headers };

    const cookies = ctx?.resp.cookies;

    if (cookies && Object.keys(cookies).length) {
      headers['Set-Cookie'] = Object.values(cookies);
    }

    return {
      headers,
    };
  },
});
```

```typescript
// context.ts
import { type inferAsyncReturnType } from '@trpc/server';
import cookie from 'cookie';

type Response = {
  headers: Record<string, string>;
  cookies: Record<string, string>;
  cookie(name: string, value: string, options?: cookie.CookieSerializeOptions): void;
  clearCookie(name: string, options?: cookie.CookieSerializeOptions): void;
  header(key: string, value: string): void;
};

export const createContext = async () => {
  const resp: Response = {
    headers: {},
    header(key: string, value: string) {
      this.headers[key] = value;
    },
    cookies: {},
    cookie(name, val, options) {
      this.cookies[name] = cookie.serialize(name, val, options);
    },
    clearCookie(name, options = {}) {
      this.cookies[name] = cookie.serialize(name, '', { ...options, expires: new Date(0) });
    },
  };

  return {
    resp,
  };
};

export type Context = inferAsyncReturnType<typeof createContext>;
```
